src/crimson/osd/osd_operations/pg_splitting.cc

```diff
@@ -0,0 +1,150 @@
+      << " to=" << next_epoch;
+  lhs << ")";
+}
+
+void PGSplitting::dump_detail(Formatter *f) const
+{
+    f->open_object_section("PGSplitting");
+    f->dump_stream("pgid") << pg->get_pgid();
+    f->dump_int("from", cur_map->get_epoch());
+    f->dump_int("to", next_epoch);
+    f->close_section();
+}
+
+
+seastar::future<> PGSplitting::start()
+{
+  logger().debug("{}: start", *this);
+  unsigned old_pg_num;
+  using cached_map_t = OSDMapService::cached_map_t;
+
+  if (cur_map->have_pg_pool(pg->get_pgid().pool())) {
+    old_pg_num = cur_map->get_pg_num(pg->get_pgid().pool());
+  } else {
+    logger().debug("{} pool doesn't exist in {}", pg->get_pgid(), cur_map);
+    return seastar::now();
+  }
+  
+  return shard_services.get_map(next_epoch).then(
+    [this, old_pg_num] (cached_map_t&& next_map) {
+    unsigned new_pg_num = next_map->get_pg_num(pg->get_pgid().pool());
+    if (new_pg_num && new_pg_num > old_pg_num) {
+      logger().debug(" New PGs: {} and Old PGs: {}", new_pg_num, old_pg_num);
+      logger().debug(" Checking for PG splitting ");
+      std::set<spg_t> children;
+      if (pg->get_pgid().is_split(old_pg_num, new_pg_num, &children)) {
+        logger().debug(" {} PG has split ", pg->get_pgid());
+        logger().debug(" {} has {} children ", pg->get_pgid(), children.size());
+        std::set<Ref<PG>> split_pgs;
+        return seastar::do_with(pg, children,
+	                        std::move(next_map),
+				new_pg_num,
+				split_pgs, [this] (auto &pg,
+						   auto &children,
+						   auto &next_map,
+						   auto &new_pg_num,
+						   auto &split_pgs) {
+	  return seastar::do_for_each(children, [this, &pg, &next_map, &new_pg_num, &split_pgs] (auto& child_pg) {
+	    return shard_services.get_or_create_pg(child_pg).then([child_pg]
+		 (auto core) {
+		 logger().debug(" PG {} mapped to {}", child_pg.pgid, core);
+		 return seastar::now();
+	       }).then([this, child_pg, next_map=next_map] {
+		 logger().debug(" {} map epoch: {}", child_pg.pgid, next_map->get_epoch());
+		 return shard_services.make_pg(std::move(next_map), child_pg, true);
+	       }).then([this, &pg, &new_pg_num, &split_pgs] (Ref<PG> child_pg) {
+		 logger().debug(" Parent PG: {}", pg->get_pgid());
+		 logger().debug(" Child PG ID: {}", child_pg->get_pgid());
+
+		 const coll_t cid{child_pg->get_pgid()};
+		 unsigned split_bits = child_pg->get_pgid().get_split_bits(new_pg_num);
+		 logger().debug(" pg num is {}, m_seed is {}, split bits is {}", new_pg_num, child_pg->get_pgid().ps(), split_bits);
+		 return pg->split_colls(child_pg->get_pgid(), split_bits, child_pg->get_pgid().ps(),
+					&child_pg->get_pgpool().info, rctx.transaction).then(
+					[this, &pg, child_pg=std::move(child_pg), split_bits, &split_pgs] () {
+					logger().debug(" {} split collection done", child_pg->get_pgid());
+					pg->split_into(child_pg->get_pgid().pgid, child_pg, split_bits);
+					split_pgs.insert(child_pg);
+				      });
+		 });
```

Comment:
This can be a good opportunity to:
* Use helper functions to make start() more friendly
* Use coroutines instead of chaining futures with `then`. As most of the lambdas here are seastar::now() this should be quite smooth. You can grep for co_await for other existing usages.